extmod/network_ninaw10.c

```diff
@@ -53,13 +54,61 @@ typedef struct _nina_obj_t {
 } nina_obj_t;
 
 // For auto-binding UDP sockets
-#define BIND_PORT_RANGE_MIN (65000)
-#define BIND_PORT_RANGE_MAX (65535)
+#define BIND_PORT_RANGE_MIN     (65000)
+#define BIND_PORT_RANGE_MAX     (65535)
+
+#define SOCKET_IOCTL_FIONREAD   (0x4004667F)
+#define SOCKET_IOCTL_FIONBIO    (0x8004667E)
+
+#define SOCKET_POLL_RD          (0x01)
+#define SOCKET_POLL_WR          (0x02)
+#define SOCKET_POLL_ERR         (0x04)
+
+#define SO_ACCEPTCONN           (0x0002)
+#define SO_ERROR                (0x1007)
+#define SO_TYPE                 (0x1008)
+#define SO_NO_CHECK             (0x100a)
+
+#define is_nonblocking_error(errno) ((errno) == MP_EAGAIN || (errno) == MP_EWOULDBLOCK || (errno) == MP_EINPROGRESS)
+
+#define debug_printf(...) // mp_printf(&mp_plat_print, __VA_ARGS__)
 
 static uint16_t bind_port = BIND_PORT_RANGE_MIN;
 const mod_network_nic_type_t mod_network_nic_type_nina;
 static nina_obj_t network_nina_wl_sta = {{(mp_obj_type_t *)&mod_network_nic_type_nina}, false, MOD_NETWORK_STA_IF};
 static nina_obj_t network_nina_wl_ap = {{(mp_obj_type_t *)&mod_network_nic_type_nina}, false, MOD_NETWORK_AP_IF};
+static bool poll_scheduled = false;
+
+STATIC mp_obj_t network_ninaw10_poll_sockets(mp_obj_t none_in) {
+    for (mp_uint_t i = 0; i < MP_STATE_PORT(mp_wifi_sockpoll_list)->len; i++) {
+        mod_network_socket_obj_t *socket = MP_STATE_PORT(mp_wifi_sockpoll_list)->items[i];
+        uint8_t flags = 0;
+        if (socket->callback == MP_OBJ_NULL || nina_socket_poll(socket->fileno, &flags) < 0) {
+            // remove from poll list on error.
+            socket->callback = MP_OBJ_NULL;
+            mp_obj_list_remove(MP_STATE_PORT(mp_wifi_sockpoll_list), socket);
+        } else if (flags && mp_sched_schedule(socket->callback, MP_OBJ_FROM_PTR(socket))) {
```

Comment:
This function is already being executed in `mp_sched_schedule` context, so you just call `socket->callback` directly, right here (eg using `mp_call_function_1`).  There shouldn't be any downside to that, and it means the latency is much smaller for the callback being executed.